x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/SparseVectorQueryBuilder.java

```diff
@@ -0,0 +1,369 @@
+        }
+
+        if (modelId != null) {
+            CoordinatedInferenceAction.Request inferRequest = CoordinatedInferenceAction.Request.forTextInput(
+                modelId,
+                List.of(modelText),
+                TextExpansionConfigUpdate.EMPTY_UPDATE,
+                false,
+                InferModelAction.Request.DEFAULT_TIMEOUT_FOR_API
+            );
+            inferRequest.setHighPriority(true);
+            inferRequest.setPrefixType(TrainedModelPrefixStrings.PrefixType.SEARCH);
+
+            SetOnce<TextExpansionResults> textExpansionResultsSupplier = new SetOnce<>();
+            queryRewriteContext.registerAsyncAction(
+                (client, listener) -> executeAsyncWithOrigin(
+                    client,
+                    ML_ORIGIN,
+                    CoordinatedInferenceAction.INSTANCE,
+                    inferRequest,
+                    ActionListener.wrap(inferenceResponse -> {
+
+                        if (inferenceResponse.getInferenceResults().isEmpty()) {
+                            listener.onFailure(new IllegalStateException("inference response contain no results"));
+                            return;
+                        }
+
+                        if (inferenceResponse.getInferenceResults().get(0) instanceof TextExpansionResults textExpansionResults) {
+                            textExpansionResultsSupplier.set(textExpansionResults);
+                            listener.onResponse(null);
+                        } else if (inferenceResponse.getInferenceResults().get(0) instanceof WarningInferenceResults warning) {
+                            listener.onFailure(new IllegalStateException(warning.getWarning()));
+                        } else {
+                            listener.onFailure(
+                                new IllegalStateException(
+                                    "expected a result of type ["
+                                        + TextExpansionResults.NAME
+                                        + "] received ["
+                                        + inferenceResponse.getInferenceResults().get(0).getWriteableName()
+                                        + "]. Is ["
+                                        + modelId
+                                        + "] a compatible model?"
+                                )
+                            );
+                        }
+                    }, listener::onFailure)
+                )
+            );
+
+            return new SparseVectorQueryBuilder(this, textExpansionResultsSupplier);
+
+        } else {
+            return vectorDimensionsToQuery(fieldName, vectorDimensions);
+        }
+
+    }
+
+    private QueryBuilder vectorDimensionsToQuery(String fieldName, List<VectorDimension> vectorDimensions) {
+        if (tokenPruningConfig != null) {
+            VectorDimensionsQueryBuilder vectorDimensionsQueryBuilder = new VectorDimensionsQueryBuilder(
```

Comment:
I don't think we need a `VectorDimensionsQueryBuilder` at all.

Reply:
The semantic query has a dependency we need to consider here. When using an ELSER-based inference service, queries against `semantic_text` fields will need to be rewritten to the equivalent of `VectorDimensionsQueryBuilder`. Instead of duplicating this query logic in the semantic query, it's probably better to use `VectorDimensionsQueryBuilder` to abstract away these query details.

Another factor is that currently, `VectorDimensionsQueryBuilder` is in the `ml` plugin. We would need to move it to a more common place so that the semantic query can access it.

I see two ways to resolve this:

- Remove `VectorDimensionsQueryBuilder` and move `SparseVectorQueryBuilder` to a common location (such as the `core` plugin). We could then use `SparseVectorQueryBuilder` from the semantic query. Is this OK from a licensing POV though? I think the ML plugin requires a platinum license... (could be wrong, that understanding is based on a cursory glance of the ML plugin config)
- Keep `VectorDimensionsQueryBuilder` and move it to a common location.

I'm fine with either, no strong preference.